cpp/src/community/detail/mis_impl.cuh

```diff
@@ -0,0 +1,793 @@
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+rmm::device_uvector<vertex_t> compute_mis(
+  raft::handle_t const& handle,
+  cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+  std::optional<cugraph::edge_property_view_t<edge_t, weight_t const*>> edge_weight_view)
+{
+  using GraphViewType = cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu>;
+  using FlagType      = vertex_t;
+
+  vertex_t local_vtx_partitoin_size = graph_view.local_vertex_partition_range_size();
+
+  bool debug = graph_view.local_vertex_partition_range_size() < 40;
+  if (debug) {
+    auto offsets = graph_view.local_edge_partition_view(0).offsets();
+    auto indices = graph_view.local_edge_partition_view(0).indices();
+    cudaDeviceSynchronize();
+    std::cout << "---- MIS Graph -----" << std::endl;
+    raft::print_device_vector("offsets: ", offsets.data(), offsets.size(), std::cout);
+    raft::print_device_vector("indices: ", indices.data(), indices.size(), std::cout);
+  }
+
+  rmm::device_uvector<vertex_t> remaining_vertices(local_vtx_partitoin_size, handle.get_stream());
+
+  auto vertex_begin =
+    thrust::make_counting_iterator(graph_view.local_vertex_partition_range_first());
+  auto vertex_end = thrust::make_counting_iterator(graph_view.local_vertex_partition_range_last());
+
+  // Compute out-degree and ranks
+  auto out_degrees = graph_view.compute_out_degrees(handle);
+
+  // Vertices with non-zero outdegree are possible candidates for MIS.
+  remaining_vertices.resize(
+    thrust::distance(remaining_vertices.begin(),
+                     thrust::copy_if(handle.get_thrust_policy(),
+                                     vertex_begin,
+                                     vertex_end,
+                                     out_degrees.begin(),
+                                     remaining_vertices.begin(),
+                                     [] __device__(auto deg) { return deg > 0; })),
+    handle.get_stream());
+
+  if (debug) {
+    cudaDeviceSynchronize();
+    raft::print_device_vector("degrees: ", out_degrees.data(), out_degrees.size(), std::cout);
+    raft::print_device_vector(
+      "remaining_vertices: ", remaining_vertices.data(), remaining_vertices.size(), std::cout);
+  }
+
+  // Each vertex's ID is its rank
+  rmm::device_uvector<vertex_t> ranks(local_vtx_partitoin_size, handle.get_stream());
+  thrust::copy(handle.get_thrust_policy(), vertex_begin, vertex_end, ranks.begin());
+
+  thrust::for_each(
+    handle.get_thrust_policy(),
+    vertex_begin,
+    vertex_end,
+    [out_degrees = raft::device_span<edge_t const>(out_degrees.data(), out_degrees.size()),
+     ranks       = raft::device_span<vertex_t>(ranks.data(), ranks.size()),
+     v_first     = graph_view.local_vertex_partition_range_first()] __device__(auto v) {
+      auto v_offset = v - v_first;
+
+      if (out_degrees[v_offset] == 0) { ranks[v_offset] = invalid_vertex_id<vertex_t>::value; }
```

Comment:
This won't work if vertex_t is unsigned. Note that invalid_vertex_id = -1 if signed and max() if unsigned.

Reply:
And you can use thrust::replace_if.

https://thrust.github.io/doc/group__replacing_ga9953891234efd44b09942f7d6381ff79.html#ga9953891234efd44b09942f7d6381ff79